platform/engineering/design-docs/2020-08-11-redirects-from-cms.md

```diff
@@ -0,0 +1,224 @@
+
+### Guiding Principles
+1. Veterans should not get 404 errors.
+2. Content on VA.gov should not have broken links to content also on VA.gov.
+3. As the CMS scales from ~20 editors and daily content releases, up to ~500 editors and continuous content releases, we can not allow a broken link to prevent content releases.
+4. Developers currently waste far too much time chasing down broken links.
+
+
+### High Level Design
+
+Three parts of the system work together:
+
+#### I. CMS
+At the highest level, the CMS manages redirects already.  It tracks them, sorts out and prevents circular references (infinite redirects), creates new redirects for moved pages and can handle query parameters, if needed for legacy handling. The CMS even provides a nice user interface for adding, deleting, and editing source, destination and status codes.  The CMS currently makes redirects available to both GraphQL and the static file content exports.  Only one minor change is needed on the CMS side to translate node/# paths to the proper alias.  We simply need a way for the Front-End to act on the redirects in the CMS. The CMS treats redirects like content, and the Front-End should too.  When a title changes, in most cases, the URL changes, so a redirect is needed.  Any Front-End system that does not account for that is a 404 factory, which is our current condition.
+
+#### II. Vets-website content creation.
+Local environments that are not AWS based would need to function correctly with automated tests and human users.  So one part of this strategy is to create client side redirects by creating an html file with a meta refresh tag
+`<meta http-equiv="refresh" content="0; URL='/new/path/destination'" />` for each redirect object found in the content export.  This will happen at the same time that content html pages are built.
+
+#### III. Vets-website deploy to AWS
+During the vets-website deploy to AWS S3, the system currently generates an object for each static page of the VA.gov website.  Via the same object creation process we can add an [S3 object with redirect metadata](https://docs.aws.amazon.com/AmazonS3/latest/dev/how-to-page-redirect.html) that describes both the destination and the return code for each HTML page that contains a meta-refresh tag. These S3 objects get discarded and rebuilt on each content release, so they would **always** be in sync with the content from the CMS.  By adding these, the website when running in an AWS environment would have the more desireable server side redirects instead of the browser-based `meta-refresh` redirects.
+
+## Specifics
+
+### Detailed Design
+
+#### I. CMS
+The redirects are output in the content export like this:
```

Comment:
this format seems to assume that all paths (source and destination) are simply paths on the same domain, but my understanding is that we handle redirects across VA.gov domains. can you speak to that?

Reply:
Those would still be handled wherever they are now  (either at the nginx or F5 level.

Reply:
Would like to still have a support capability within Drupal to find all links and update them when a URL outside the CMS is changed.